sdk/eventhub/Azure.Messaging.EventHubs/samples/Program.cs

```diff
@@ -42,27 +52,172 @@ public static async Task Main(string[] args)
-                parsedArgs.EventHub = Console.ReadLine().Trim();
+                PromptConnectionString(parsedArgs);
+                PromptEventHubName(parsedArgs);
+
+                await eventHubsSample.RunAsync(parsedArgs.ConnectionString, parsedArgs.EventHub);
+            }
+            else if (sample is IEventHubsIdentitySample identitySample)
+            {
+                PromptFullyQualifiedNamespace(parsedArgs);
+                PromptEventHubName(parsedArgs);
+                PromptTenantId(parsedArgs);
+                PromptClientId(parsedArgs);
+                PromptSecret(parsedArgs);
+
+                await identitySample.RunAsync(parsedArgs.FullyQualifiedNamespace,
+                                              parsedArgs.EventHub,
+                                              parsedArgs.TenantId,
+                                              parsedArgs.ClientId,
+                                              parsedArgs.Secret);
+            }
+        }
+
+        /// <summary>
+        ///   It prints a list of options to console for the user. It reads the user's
+        ///   choice then it returns the chosen sample.
+        /// </summary>
+        ///
+        /// <returns>A sample or null if that could not be determined.</returns>
+        ///
+        private static INamedSample RetrieveSample()
+        {
+            var samples = LocateSamples();
+
+            PrintEventHubsSamples(samples);
+
+            var choice = ReadSelection(samples);
+
+            if (IsEventHubsIdentity(samples, choice))
+            {
+                var identitySamples = LocateIdentitySamples();
+
+                PrintEventHubsIdentitySamples(identitySamples);
+
+                while (IsEventHubsIdentity(samples, choice))
+                {
+                    choice = ReadSelection(identitySamples);
+
+                    if (choice.HasValue && !IsGoBack(identitySamples, choice))
+                    {
+                        return identitySamples[choice.Value];
+                    }
+                    else if (IsGoBack(identitySamples, choice))
+                    {
+                        choice = ReadSelection(samples);
+                    }
+                }
+            }
+
+            if (choice.HasValue)
+            {
+                return samples[choice.Value];
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        ///   It checks if an option is to go back.
+        /// </summary>
+        ///
+        /// <param name="samples">A list of identity samples</param>
+        /// <param name="choice">The choice made from console</param>
+        ///
+        /// <returns>If the user has chosed to go back in the main sample listing.</returns>
+        ///
+        private static bool IsGoBack(IReadOnlyList<IEventHubsIdentitySample> identitySamples, int? choice)
```

Comment:
You allow for a set of extra options, but then make a positional assumption based on the last option.    Should we either make this a text-based comparison or make the "extra options" a single member to avoid mistakes... or am I reading this wrong and missing a detail?

Reply:
Yes, there is an important very "sneaky" detail here. My fault here if it was not evident my apologies for missing the required level of detail.

The indexing is zero-based and the right implementation would be:

```csharp

/// <summary>
///   It checks if an option is to go back.
/// </summary>
///
/// <param name="samples">A list of identity samples</param>
/// <param name="choice">The zero based index referring to the option chosen from console</param>
///
/// <returns>If the user has chosed to go back in the main sample listing.</returns>
///
private static bool IsGoBack(IReadOnlyList<IEventHubsIdentitySample> identitySamples, int? choice)
{
   return IsLastOption(identitySamples, choice);
}

private static bool IsLastOption(IReadOnlyList<IEventHubsIdentitySample> identitySamples, int? choice)
{
    return choice == (identitySamples.Count + extraOptionsForIdentitySamples.Count() - 1);
}

```

I needed to split the extra options into two different lists so to have to compute the maximum option allowed for each category.

Reply:
I take you meant ```Length``` rather than ```Count```

Reply:
Yup.  My bad.   I read that as the list coming in.   Sorry about that.

Reply:
No worries mate, you did great